extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -0,0 +1,346 @@
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.RowSignature;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class WindowOperatorQueryKit implements QueryKit<WindowOperatorQuery>
+{
+  private final ObjectMapper jsonMapper;
+  public static final String CTX_SCAN_SIGNATURE = "scanSignature";
+
+  public WindowOperatorQueryKit(ObjectMapper jsonMapper)
+  {
+    this.jsonMapper = jsonMapper;
+  }
+
+  @Override
+  public QueryDefinition makeQueryDefinition(
+      String queryId,
+      WindowOperatorQuery originalQuery,
+      QueryKit<Query<?>> queryKit,
+      ShuffleSpecFactory resultShuffleSpecFactory,
+      int maxWorkerCount,
+      int minStageNumber
+  )
+  {
+    // need to validate query first
+    // populate the group of operators to be processed as each stage
+    // the size of the operators is the number of serialized stages
+    // later we should also check if these can be parallelized
+    // check there is an empty over clause or not
+    List<List<OperatorFactory>> operatorList = new ArrayList<>();
+    boolean status = validateAndReturnOperatorList(originalQuery, operatorList);
+
+
+    ShuffleSpec nextShuffleSpec = findShuffleSpecForNextWindow(operatorList.get(0), maxWorkerCount);
+    // add this shuffle spec to the last stage of the inner query
+
+    final QueryDefinitionBuilder queryDefBuilder = QueryDefinition.builder().queryId(queryId);
+    if (nextShuffleSpec != null) {
+      final ClusterBy windowClusterBy = nextShuffleSpec.clusterBy();
+      originalQuery = (WindowOperatorQuery) originalQuery.withOverriddenContext(ImmutableMap.of(
+          DataSourcePlan.NEXT_SHUFFLE_COL,
+          windowClusterBy
+      ));
+    } else {
+      nextShuffleSpec = ShuffleSpecFactories.singlePartition()
+                                            .build(ClusterBy.none(), false);
+    }
+    final DataSourcePlan dataSourcePlan = DataSourcePlan.forDataSource(
+        queryKit,
+        queryId,
+        originalQuery.context(),
+        originalQuery.getDataSource(),
+        originalQuery.getQuerySegmentSpec(),
+        originalQuery.getFilter(),
+        null,
+        maxWorkerCount,
+        minStageNumber,
+        false
+    );
+
+    // Handling for case where leaf operators are present
+    // leaf operators can only be a scan,
+    // and they execute before any other operators,
+    // so we introduce a separate stage for the leaf operators.
+    // We recreate scan query from leaf operator and use the associated
+    // ScanQueryFrameProcessorFactory for that particular stage.
+    // The next stage will work on the output of this stage
+    if (!originalQuery.getLeafOperators().isEmpty()) {
+      final RowSignature originalSign = originalQuery.getRowSignature();
+      for (OperatorFactory of : originalQuery.getLeafOperators()) {
```

Comment:
This part of leaf operators still needs work:

1. It has only access to column names and need the column types which are not available
2. Since this runs at the controller, we cannot get the storage adapter. Also it is illogical to get the column type by going over all segments
3. Probably a better way is to include the data types in the leaf operator at the time of planning and pass through to use here

Reply:
Addressed with point no 6. mentioned in the description. For MSQ we ensure the query is planned as a window over an inner scan without the leafOperator and hence they are planned as a windowing stage after a scan stage